packages/plugin-core/src/commands/CreateSchemaFromHierarchyCommand.ts

```diff
@@ -0,0 +1,528 @@
+    ancestorCandidate.note.fname + "."
+  );
+  return isChild;
+}
+
+function createCandidatesMapByFname(items: readonly SchemaCandidate[]) {
+  return new Map(items.map((item) => [item.note.fname, item]));
+}
+
+function getUriFromSchema(schema: SchemaModuleProps) {
+  const vaultPath = vault2Path({
+    vault: schema.vault,
+    wsRoot: getDWorkspace().wsRoot,
+  });
+  const uri = Uri.file(
+    SchemaUtils.getPath({ root: vaultPath, fname: schema.fname })
+  );
+  return uri;
+}
+
+function getSchemaUri(vault: DVault, schemaName: string) {
+  const vaultPath = vault2Path({ vault, wsRoot: getDWorkspace().wsRoot });
+  const uri = Uri.file(
+    SchemaUtils.getPath({ root: vaultPath, fname: schemaName })
+  );
+  return uri;
+}
+
+/**
+ * Encapsulates methods that are responsible for user interaction when
+ * asking user for input data.
+ * */
+export class UserQueries {
+  static async haveUserPickSchemaFileName(
+    hierarchyLevel: HierarchyLevel,
+    vault: DVault
+  ): Promise<string | undefined> {
+    let alreadyExists = false;
+    let schemaName: string | undefined;
+
+    do {
+      // eslint-disable-next-line no-await-in-loop
+      schemaName = await VSCodeUtils.showInputBox({
+        value: hierarchyLevel.getDefaultSchemaName(),
+      });
+
+      if (!schemaName) {
+        // Cancelled.
+        return schemaName;
+      }
+
+      alreadyExists = fs.existsSync(getSchemaUri(vault, schemaName).fsPath);
+      if (alreadyExists) {
+        vscode.window.showInformationMessage(
+          `Schema with name '${schemaName}' already exists. Pick a different name or cancel.`
+        );
+      }
+    } while (alreadyExists);
+
+    return schemaName;
+  }
+
+  static async haveUserSelectHierarchyLevel(currDocFsPath: string) {
+    const hierarchy = new Hierarchy(path.basename(currDocFsPath, ".md"));
+
+    if (hierarchy.depth() <= 2) {
+      // We require some depth to the hierarchy to be able to choose a variance
+      // pattern within in it. More info within Hierarchy object.
+      await vscode.window.showErrorMessage(
+        `Pick a note deeper within the note hierarchy.`
+      );
+
+      return undefined;
+    }
+
+    if (PluginSchemaUtils.doesSchemaExist(hierarchy.topId())) {
+      // To avoid unpredictable conflicts of schemas: for now we will not allow
+      // creation schemas for hierarchies that already have existing top
+      // level schema id. Instead we will pop up error message with navigation
+      // action to the existing schema.
+      const msgGoToSchema = "Go to schema";
+      const action = await vscode.window.showErrorMessage(
+        `Schema with top level id: '${hierarchy.topId()}' already exists.`,
+        msgGoToSchema
+      );
+      if (action === msgGoToSchema) {
+        const schema = PluginSchemaUtils.getSchema(hierarchy.topId());
+
+        await VSCodeUtils.openFileInEditor(getUriFromSchema(schema));
+      }
+
+      return undefined;
```

Comment:
It'd be good if we can capture some of these exit points (`return undefined`) with some analytics in a finally block.  That way we can distinguish between a user who ran the command but gave up vs a user who went all the way through and ended up with a schema. This method should allow you to add props to the analytics payload that gets generated from the command invocation: https://github.com/dendronhq/dendron/blob/01d2dd8b69d6bff175acf78405bd903ac268b3b3/packages/plugin-core/src/commands/base.ts#L32

Reply:
Agreed, as discussed will circle back to this.